aws/resource_aws_appsync_datasource_test.go

```diff
@@ -0,0 +1,439 @@
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAwsAppsyncDatasourceDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAppsyncDatasourceConfig_ddb(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsAppsyncDatasourceExists("aws_appsync_datasource.test"),
+					resource.TestCheckResourceAttrSet("aws_appsync_datasource.test", "arn"),
+					resource.TestCheckResourceAttr("aws_appsync_datasource.test", "type", "AMAZON_DYNAMODB"),
+				),
+			},
+			{
+				Config: testAccAppsyncDatasourceConfig_update_lambda(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsAppsyncDatasourceExists("aws_appsync_datasource.test"),
+					resource.TestCheckResourceAttrSet("aws_appsync_datasource.test", "arn"),
+					resource.TestCheckResourceAttr("aws_appsync_datasource.test", "type", "AWS_LAMBDA"),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckAwsAppsyncDatasourceDestroy(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*AWSClient).appsyncconn
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_appsync_datasource" {
+			continue
+		}
+
+		input := &appsync.GetDataSourceInput{
+			ApiId: aws.String(rs.Primary.Attributes["api_id"]),
+			Name:  aws.String(rs.Primary.Attributes["name"]),
+		}
+
+		_, err := conn.GetDataSource(input)
+		if err != nil {
+			if isAWSErr(err, appsync.ErrCodeNotFoundException, "") {
+				return nil
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func testAccCheckAwsAppsyncDatasourceExists(name string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		_, ok := s.RootModule().Resources[name]
+		if !ok {
+			return fmt.Errorf("Not found: %s", name)
+		}
+
+		return nil
+	}
+}
+
+func testAccAppsyncDatasourceConfig_ddb(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_appsync_graphql_api" "test" {
+  authentication_type = "API_KEY"
+  name = "tf_appsync_%s"
+}
+
+resource "aws_dynamodb_table" "test" {
+  name = "tf-ddb-%s"
+  read_capacity  = 1
+  write_capacity = 1
+  hash_key = "UserId"
+  attribute {
+    name = "UserId"
+    type = "S"
+  }
+}
+
+resource "aws_iam_role" "test" {
+  name = "tf-role-%s"
+
+  assume_role_policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": "sts:AssumeRole",
+      "Principal": {
+        "Service": "appsync.amazonaws.com"
+      },
+      "Effect": "Allow"
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_iam_role_policy" "test" {
+  name = "tf-rolepolicy-%s"
+  role = "${aws_iam_role.test.id}"
+
+  policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": [
+        "dynamodb:*"
+      ],
+      "Effect": "Allow",
+      "Resource": [
+        "${aws_dynamodb_table.test.arn}"
+      ]
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_appsync_datasource" "test" {
+  api_id = "${aws_appsync_graphql_api.test.id}"
+  name = "tf_appsync_%s"
+  type = "AMAZON_DYNAMODB"
+  dynamodb_config {
+    region = "us-west-2"
```

Comment:
Do you mind swapping this for a data source, so the test is usable in any region?
https://www.terraform.io/docs/providers/aws/d/region.html